### PR TITLE
[Heartbeat]: limit parallelization of tasks by jobtype

### LINKED ISSUE #23216: [Heartbeat] Limit number of parallel monitors, by type
Today, heartbeat has the `heartbeat.scheduler.limit` option, which limits the total number of internal tasks heartbeat can run. For lightweight checks this corresponds to the number of open file descriptors.

This doesn't work well for synthetics, where each check is made of a single recursive run of many tasks (one per JSON line from the agent) that don't correlate to FDs at all.

We should add a new setting:

`heartbeat.jobs.{icmp|http|tcp|browser}.limit` which lets users define on a job-type basis how many concurrent tasks may be run.

Docs are updated to clarify how these relate and should be used

### PR DESCRIPTION
**Pull request by vigneshshanmugam**
+ fixes #23216

### CODE REVIEW

File: heartbeat/scheduler/scheduler_test.go [hunk 1]
```diff
@@ -208,7 +210,10 @@ func TestScheduler_runRecursiveTask(t *testing.T) {
 	for _, testCase := range testCases {
 		t.Run(testCase.name, func(t *testing.T) {
 			limit := int64(100)
-			s := NewWithLocation(limit, monitoring.NewRegistry(), tarawaTime())
+			s := NewWithLocation(limit, monitoring.NewRegistry(), tarawaTime(), nil)
+
+			jobSem := semaphore.NewWeighted(math.MaxInt64)
+			jobSem.Acquire(context.Background(), 1)
```
Comment (vigneshshanmugam) [inline heartbeat/scheduler/scheduler_test.go:216]: needed to acquire as the semaphore acquire is handled by `runRecursiveJob` in main scheduler.

File: heartbeat/scheduler/scheduler.go [hunk 2]
```diff
@@ -77,13 +79,33 @@ type Schedule interface {
 	RunOnInit() bool
 }
 
+func getJobLimitSem(jobLimitByType map[string]config.JobLimit) map[string]*semaphore.Weighted {
+	jc := map[string]config.JobLimit{
```
Comment (andrewvc) [inline heartbeat/scheduler/scheduler.go:83]: I think we can remove this, we don't need limits for non-configured types.

If there's a missing limit we can just make `Acquire` and `Release` noops, which already appears to be the case.

File: heartbeat/scheduler/scheduler_test.go (hunk 1 above)
Comment (andrewvc) [reply heartbeat/scheduler/scheduler_test.go:217]: Can we make this an inline comment? Future coders will be grateful!

**Review by andrewvc — CHANGES_REQUESTED** [review]
Looks great, needs a couple minor changes

File: heartbeat/scheduler/scheduler_test.go (hunk 1 above)
Comment (vigneshshanmugam) [reply heartbeat/scheduler/scheduler_test.go]: Removed that code path.

File: heartbeat/scheduler/scheduler.go (hunk 2 above)
Comment (vigneshshanmugam) [reply heartbeat/scheduler/scheduler.go]: Done.

File: heartbeat/scheduler/scheduler_test.go [hunk 3]
```diff
@@ -240,8 +242,76 @@ func TestScheduler_runRecursiveTask(t *testing.T) {
...
+		callback()
+		if num < 1 {
+			return nil
+		}
+		return []TaskFunc{makeTasks(num-1, callback)}
+	}
+}
+
+func TestScheduler_runRecursiveJob(t *testing.T) {
+	tests := []struct {
```
Comment (andrewvc) [inline heartbeat/scheduler/scheduler_test.go:256]: Can we add one last test for running jobs when no limit is specified?

**Review by andrewvc — CHANGES_REQUESTED** [review]

File: heartbeat/scheduler/scheduler_test.go (hunk 3 above)
Comment (vigneshshanmugam) [reply heartbeat/scheduler/scheduler_test.go:256]: Done.

**Review by andrewvc — APPROVED** [review]
LGTM

**Pull request merged**